src/blockdev.rs

```diff
@@ -461,6 +464,205 @@ impl Mount {
+                    "sector size {} of disk {} doesn't match sector size {} of saved GPT",
+                    disk_sector_size,
+                    disk.display(),
+                    self.sector_size.unwrap()
+                );
+            }
+        }
+        let mut gpt = match GPT::find_from(&mut f) {
+            Ok(gpt) => gpt,
+            Err(e) => {
+                // Bad or missing GPT.  Maybe the caller just deleted it.
+                // Create a new partition table and save to that, otherwise
+                // we risk data loss.  If the error is anything fancier than
+                // a missing table, warn about it.
+                match e {
+                    gptman::Error::InvalidSignature => (),
+                    _ => {
+                        eprintln!("Couldn't read partition table: {}.  Recreating.", e);
+                    }
+                }
+                GPT::new_from(
+                    &mut f,
+                    self.sector_size.expect("sector size unset"),
+                    *Uuid::new_v4().as_bytes(),
+                )
+                .chain_err(|| format!("creating new partition table for {}", disk.display()))?
+            }
+        };
+        if gpt.sector_size != self.sector_size.expect("sector size unset") {
+            // install will fail on an image that doesn't match the disk,
+            // so this shouldn't happen
+            bail!(
+                "sector size {} of GPT on {} doesn't match sector size {} of saved GPT",
+                gpt.sector_size,
+                disk.display(),
+                self.sector_size.unwrap()
+            );
+        }
+
+        // Fail if the last on-disk partition overlaps with the beginning of
+        // the first saved partition.  Ignore holes.  This test is distinct
+        // from the download-time LimitReader checking, because the image
+        // may claim to have partitions beyond the end of the image file.
+        // If this occurs, install() will restore the saved partitions after
+        // clearing the table.
+        if let Some((i_end, end)) = gpt.iter().max_by_key(|(_, p)| p.ending_lba) {
+            if let Some((i_start, start)) =
+                self.partitions.iter().min_by_key(|(_, p)| p.starting_lba)
+            {
+                if end.ending_lba >= start.starting_lba {
+                    bail!(
+                        "disk partition {} ('{}') ends after start of saved partition {} ('{}')",
+                        i_end,
+                        end.partition_name.as_str(),
+                        i_start,
+                        start.partition_name.as_str()
+                    )
+                }
+            }
+        }
+
+        // merge saved partitions into partition table
+        // find partition number one larger than the largest used one
+        let mut next = gpt
+            .iter()
+            .fold(1, |prev, (i, e)| if e.is_used() { i + 1 } else { prev });
```

Comment:
Given how we handle `.is_used()`, should we error out if the user tries to save an unused partition?

Reply:
No, it's intentionally valid to specify a filter that doesn't match anything.  That lets the user specify all relevant partitions that might be on any given target system.